c2rust-analyze/src/util.rs

```diff
@@ -123,36 +148,43 @@ pub enum Callee<'tcx> {
 
     /// core::ptr::is_null
     IsNull,
-
-    /// Some other statically-known function, including functions defined in the current crate.
-    Other {
-        def_id: DefId,
-        substs: SubstsRef<'tcx>,
-    },
 }
 
-pub fn ty_callee<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Option<Callee<'tcx>> {
-    let (did, substs) = match *ty.kind() {
-        TyKind::FnDef(did, substs) => (did, substs),
-        _ => return None,
+pub fn ty_callee<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Callee<'tcx> {
+    let is_trivial = || {
+        let is_trivial = ty.fn_sig(tcx).is_trivial(tcx);
+        eprintln!("{ty:?} is trivial: {is_trivial}");
+        is_trivial
     };
 
-    if let Some(callee) = builtin_callee(tcx, ty, did) {
-        return Some(callee);
+    match *ty.kind() {
+        ty::FnDef(did, substs) => {
+            if is_trivial() {
+                return Callee::Trivial;
+            }
+            if let Some(callee) = builtin_callee(tcx, did) {
+                return callee;
+            }
+            if tcx.def_kind(tcx.parent(did)) == DefKind::ForeignMod {
+                return Callee::UnknownDef { ty };
+            }
+            Callee::Normal {
+                def_id: did,
+                substs,
+            }
+        }
+        ty::FnPtr(..) => {
+            if is_trivial() {
+                Callee::Trivial
+            } else {
+                Callee::UnknownDef { ty }
+            }
+        }
+        _ => Callee::Trivial,
```

Comment:
>This PR also folds None::<Callee> into Callee::Trivial, as Trivial needs no special handling, which is exactly how None is treated.

I think it was incorrect for us to treat `None` that way - that case should have been a panic instead.  If we don't know what's being called, then I think the right thing to return is `UnknownDef`.

Reply:
Okay, so the fix here would just be 
```suggestion
        _ => Callee::UnknownDef { ty },
```
?

Reply:
Yes, that looks right to me

Reply:
Fixed in 1a54c3cf4a344ebb9743f3a62ced9a66571fc820.